New tools documentation (#163)
IMAP-Science-Operations-Center/imap_processing: docs/source/reference/tools/xtce_generator.rst

.. _xtce_generator:

Generating Telemetry XML with Python Script
===========================================

This document provides steps and information on how to use
`xtce_generator_template.py` script as a base for users to generate

> telemetry XML files using the TelemetryGenerator class. The script is designed to

Reviewer: ```suggestion
telemetry XML files. The script is designed to
```

simplify the process of creating telemetry definitions for various packet types.

The script is located in the `tools/xtce_generation` directory. The script is called
`xtce_generator_template.py`. The script is a ``template`` that can be modified to
generate telemetry XML files for different packet types. Your new file should be
called `xtce_generator_yourinstrument.py`.
An example of how to use the script is `xtce_generator_codice.py` which is also
located in the `tools/xtce_generation` directory.

Prerequisites
-------------

Before you begin, ensure you have the following:

- Python installed (version 3.9 or higher)

> - The required Python packages installed:

Reviewer: Instead of saying "you need these installed" we should add pathlib to poetry and include instructions for installing that:
`poetry install --extras tools`

    - pathlib

>     - tools.xtce_generation.telemetry_generator

Reviewer 2: pathlib is a part of the standard library, and I don't think you need to install telemetry_generator but rather you just use it right?

I'd suggest indicating how you do this with the poetry environments and linking to those documents as well. Something similar to this, feel free to reword as you want.

> Generating XTCEs is only done whenever packet definitions get updated, and thus it is not a part of the main processing package. To use it there are a few extra dependencies like `pandas` that you can install with `poetry install --with-extras xtce`.

Author: I like this!

How to Use
----------

Define the instrument name in the `main()` function by setting the `instrument_name`

> variable to the name of your instrument.
> 
> ::

Reviewer 2: ```suggestion
variable to the name of your instrument::
```

I think you can eliminate this? Or alternatively you should add a `.. code::` or some qualifier, the double colons feel empty hanging out there otherwise. (Also to the ones below this)

    instrument_name = "your_instrument_name"

In the code, file paths are being configured. Make sure to change the file paths to
match your instrument's file structure.

::

    current_directory = Path(__file__).parent

    module_path = f"{current_directory}/../../imap_processing"

    # This is the path of the output directory
    packet_definition_path = f"{module_path}/{instrument_name}/packet_definitions"

    # This is the path to the excel file that contains the telemetry definitions
    path_to_excel_file = f"{current_directory}/your_packet.xlsx"

> Define packet names and apIds. This is case sensitive. The packet names and apIds

Reviewer 2: ```suggestion
Define packet names and Application Process Identifiers (APIDs). This is case sensitive. The packet names and apIds
```

Maybe even make the APID a link? https://sanaregistry.org/r/space_packet_protocol_application_process_id/

> Define packet names and apIds. This is case sensitive. The packet names and apIds

Reviewer 2: What is case sensitive? The packet name?

Do the packet names need to be exactly what the tabs of the spreadsheet are? It would be helpful to clarify what this should be or if I am free to choose what I want.

must match the names and apIds in the excel file. You can use as many packet names

> and apIds as you want. The apId should not be hexagon format. It should be an integer.

Reviewer 2: ```suggestion
and apIds as you want. The APID should be an integer (not hexadecimal).
```

> and apIds as you want. The apId should not be hexagon format. It should be an integer.

Reviewer: ```suggestion
Define packet names and apIds. This is case sensitive. You can use as many packet names
and apIds as you want. The apId should not be hexagon format. It should be an integer.
```

Remove the sentence about excel files, as other projects may have different ways of defining packets and it would be confusing

Follow the format below.

::

    packets = {
        # Define packet names and associated Application IDs (apId)
        "your_packet_A": ####,
        "your_packet_B": ####,
        # ... (other packet definitions)
    }

Generating Telemetry XML Files
-------------------------------

> This code block generates telemetry XML files for the packets defined in the
> `packets` dictionary. The telemetry XML files are saved in the directory specified
> by the `packet_definition_path` variable.

Reviewer 2: ```suggestion
Once you have your xtce processing file defined, you can run it with the following command:
    python xtce_generator_codice.py
```

I'd give the explicit command to use here after the file setup is complete. We don't care as much about what the code is doing (IMO), this is more of a user-guide for how to invoke the routines.

> 

Reviewer: Is this still a WIP? It ends kind of abruptly. This section also doesn't feel very necessary - I would just move the last sentence up to the previous section (unless you're planning on adding a lot more.)

> 

Reviewer: You can look at the other documentation for an example on how to add this file to the `toctree` in the reference index. You should also make a `tools.rst` file to act as a landing page for this tools directory - it doesn't need to include anything except another toctree, but it will act as the top-level page for any other documentation we add under tools.

Author: I asked about this in the slack thread.
